File: 14_Stack & Queue-II/M146. LRU Cache.java
// LRU Cache (Important)
// Least Recently Used cache — a very popular DSA + system design interview problem.
// https://leetcode.com/problems/lru-cache/

import java.util.LinkedHashMap;

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

class LRUCache {

    private int capacity;
    private LinkedHashMap<Integer, Integer> cache;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        // accessOrder = true → maintains order by usage (LRU at start, MRU at end)
        this.cache = new LinkedHashMap<>(capacity, 0.75f, true);
    }

    public int get(int key) {
        return cache.getOrDefault(key, -1);
    }

    public void put(int key, int value) {
        if (cache.containsKey(key)) {
            cache.remove(key);
        } else if (cache.size() == capacity) {
            // Remove least recently used (first entry in LinkedHashMap)
            cache.remove(cache.keySet().iterator().next());
        }
        cache.put(key, value);
    }

    public static void main(String[] args) {
        LRUCache cache = new LRUCache(2);
        cache.put(1, 1);
        cache.put(2, 2);
        System.out.println(cache.get(1)); // returns 1
        cache.put(3, 3); // evicts key 2
        System.out.println(cache.get(2)); // returns -1 (not found)
        cache.put(4, 4); // evicts key 1
        System.out.println(cache.get(1)); // returns -1 (not found)
        System.out.println(cache.get(3)); // returns 3
        System.out.println(cache.get(4)); // returns 4
    }
}

File: JavaScript/14_Stack & Queue-II/M146. LRU Cache.js
// 146. LRU Cache
// https://leetcode.com/problems/lru-cache/

/**
 * Creates an LRU Cache (functional approach)
 * Uses Map to maintain insertion order
 * @param {number} capacity - Maximum capacity
 * @returns {Object} LRUCache operations
 */
function createLRUCache(capacity) {
    const cache = new Map();

    function get(key) {
        if (!cache.has(key)) return -1;
        
        // Move to end (most recently used)
        const value = cache.get(key);
        cache.delete(key);
        cache.set(key, value);
        return value;
    }

    function put(key, value) {
        if (cache.has(key)) {
            cache.delete(key);
        } else if (cache.size >= capacity) {
            // Remove least recently used (first item)
            const firstKey = cache.keys().next().value;
            cache.delete(firstKey);
        }
        cache.set(key, value);
    }

    return { get, put };
}

// Test cases
const lRUCache = createLRUCache(2);
lRUCache.put(1, 1);
lRUCache.put(2, 2);
console.log(lRUCache.get(1));    // 1
lRUCache.put(3, 3);              // evicts key 2
console.log(lRUCache.get(2));    // -1 (not found)
lRUCache.put(4, 4);              // evicts key 1
console.log(lRUCache.get(1));    // -1 (not found)
console.log(lRUCache.get(3));    // 3
console.log(lRUCache.get(4));    // 4
